File: 2018/puzzle4/python/solution.py
from collections import Counter

if __name__ == "__main__":
    with open('../input.txt') as data:
        box_ids = data.readlines()
        found = False
        for i, box_id in enumerate(box_ids):
            if found:
                break
            for j, comp_id in enumerate(box_ids):
                if i != j and len(box_id) == len(comp_id):
                    id1 = list(box_id.strip())
                    id2 = list(comp_id.strip())
                    diff = 0
                    diff_idx = 0
                    for k in range(len(id1)):
                        if id1[k] != id2[k]:
                            diff += 1
                            diff_idx = k

                    if diff == 1:
                        found = True
                        break

        print(f"Similar letters: {comp_id[:diff_idx]}{comp_id[diff_idx+1:]}")

File: 2018/puzzle4/js/solution.js
const fs = require('fs')

fs.readFile('../input.txt', 'utf8', (err, data) => {
    let boxIds = data.split('\n')
    boxIds.pop()
    let found = false
    let diff = 0
    let diffIdx = 0
    let checkAgainst;

    for (let i = 0; i < boxIds.length; i++) {
        if (found) break;

        for (let j = 0; j < boxIds.length; j++) {
            if (i == j && boxIds[i].length == boxIds[j].length) continue;
            let id1 = boxIds[i]
            let id2 = boxIds[j]
            diff = 0
            diffIdx = 0
            for (let k = 0; k < id1.length; k++) {
                if (id1[k] !== id2[k]) {
                    diff++;
                    diffIdx = k;
                }
            }
            if (diff == 1) {
                found = true;
                checkAgainst = id1
                break
            }
        }
    }

    console.log(diffIdx)
    console.log(`${checkAgainst.slice(0, diffIdx)}${checkAgainst.slice(diffIdx + 1)}`)
})